src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs

```diff
@@ -657,8 +673,134 @@ public override Task WriteAsync(Byte[] array, int offset, int count, Cancellatio
 
         public override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
         {
-            return StreamHelpers.ArrayPoolCopyToAsync(this, destination, bufferSize, cancellationToken: cancellationToken);
+            // Validation as base CopyToAsync would do
+            StreamHelpers.ValidateCopyToArgs(this, destination, bufferSize);
+
+            // Validation as ReadAsync would do
+            EnsureDecompressionMode();
+            if (_asyncOperations != 0) throw new InvalidOperationException(SR.InvalidBeginCall);
+            EnsureNotDisposed();
+
+            // Early check for cancellation
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled<int>(cancellationToken);
+            }
+
+            // The caller asked for a specific bufferSize, but we're going to end up potentially using two buffers:
+            // a buffer for storing the inflated output and passing along to the destination stream, and whatever
+            // buffer the source stream creates in its CopyToAsync implementation.  As such, we halve the buffer size,
+            // to split it across the two.
+            bufferSize = Math.Max(1, bufferSize / 2);
```

Comment:
> Personally I would expect a copyToAsync(bufferSize 10) to complete in one Read/Write call when there is only 10 bytes in the source stream.

Given that DeflateStream changes the amount of data as part of the transformation, you generally can't expect 10 bytes of input with a 10 byte buffer will result in a single write to the output stream.  But I get the general point.

I'll remove the halving.  We can add it later if this looks like an issue.